src/wallet/wallets.go

```diff
@@ -106,98 +74,68 @@ func backupWltFile(src, dst string) error {
-		logger.Panic(err)
-	}
-}
-
-// Add adds wallet to current wallet
-func (wlts Wallets) Add(w Wallet) error {
-	if _, dup := wlts[w.GetFilename()]; dup {
+// add add walet to current wallet
+func (wlts Wallets) add(w *Wallet) error {
+	if _, dup := wlts[w.Filename()]; dup {
 		return ErrWalletNameConflict
 	}
 
-	wlts[w.GetFilename()] = &w
+	wlts[w.Filename()] = w
 	return nil
 }
 
-// Remove wallet of specific id
-func (wlts Wallets) Remove(id string) {
+// remove wallet of specific id
+func (wlts Wallets) remove(id string) {
 	delete(wlts, id)
 }
 
-// Get returns wallet by wallet id
-func (wlts Wallets) Get(wltID string) (*Wallet, bool) {
-	if w, ok := wlts[wltID]; ok {
+// get returns wallet by wallet id
+func (wlts Wallets) get(id string) (*Wallet, bool) {
+	if w, ok := wlts[id]; ok {
 		return w, true
 	}
-	return &Wallet{}, false
+	return nil, false
 }
 
 // set sets a wallet into the map
-func (wlts Wallets) set(w Wallet) {
-	wlts[w.GetFilename()] = &w
-}
-
-// Update updates the given wallet, return error if not exist
-func (wlts Wallets) Update(wltID string, updateFunc func(Wallet) Wallet) error {
-	w, ok := wlts[wltID]
-	if !ok {
-		return ErrWalletNotExist
-	}
-
-	newWlt := updateFunc(*w)
-	wlts[wltID] = &newWlt
-	return nil
+func (wlts Wallets) set(w *Wallet) {
+	wlts[w.Filename()] = w.clone()
 }
 
 // NewAddresses creates num addresses in given wallet
-func (wlts *Wallets) NewAddresses(wltID string, num uint64) ([]cipher.Address, error) {
-	if w, ok := (*wlts)[wltID]; ok {
+func (wlts *Wallets) newAddresses(id string, num uint64) ([]cipher.Address, error) {
+	if w, ok := (*wlts)[id]; ok {
 		return w.GenerateAddresses(num)
 	}
-	return nil, fmt.Errorf("wallet: %v does not exist", wltID)
-}
-
-// Save check for name conflicts!
-// resolve conflicts for saving wallets who have different names
-func (wlts Wallets) Save(dir string) map[string]error {
-	errs := make(map[string]error)
-	for id, w := range wlts {
-		if err := w.Save(dir); err != nil {
-			errs[id] = err
-		}
-	}
-	if len(errs) == 0 {
-		return nil
-	}
-	return errs
+	return nil, fmt.Errorf("wallet: %v does not exist", id)
 }
 
-func (wlts Wallets) toReadable(f ReadableWalletCtor) []*ReadableWallet {
+// ToReadable converts Wallets to *ReadableWallet array
+func (wlts Wallets) ToReadable() []*ReadableWallet {
 	var rw []*ReadableWallet
 	for _, w := range wlts {
-		rw = append(rw, f(*w))
+		rw = append(rw, NewReadableWallet(w))
 	}
-	sort.Sort(ByTm(rw))
+
+	sort.Slice(rw, func(i int, j int) bool {
+		return rw[i].time() < rw[j].time()
+	})
 	return rw
 }
 
-// ToReadable converts Wallets to *ReadableWallet array
-func (wlts Wallets) ToReadable() []*ReadableWallet {
-	return wlts.toReadable(NewReadableWallet)
+// Update updates the given wallet, return error if not exist
+func (wlts Wallets) update(id string, updateFunc func(*Wallet) (*Wallet, error)) error {
+	w, ok := wlts[id]
+	if !ok {
+		return ErrWalletNotExist
+	}
+
+	newWlt, err := updateFunc(w.clone())
```

Comment:
We update the clone wallet, in case of interrupting the original wallet in the updateFunc. Actually the returned `newWlt` is the cloned wallet, and will be used to replace the old wallet, therefore, call erase of the cloned wallet would make the wallet lose seeds, address keys, etc. 

I will change the signature of `updateFunc`, it shouldn't have *Wallet returned, which might confuse user, and don't know if the input and returned wallet are related.